tools/probes/probes_main.c

```diff
@@ -0,0 +1,288 @@
+
+	files[i].fd = fopen(path, "wb");
+	if (!files[i].fd) {
+		fprintf(stderr, "error: unable to create file %s, error %d\n",
+			path, errno);
+		exit(0);
+	}
+
+	files[i].buffer_id = buffer_id;
+
+	files[i].header.riff.chunk_id = HEADER_RIFF;
+	files[i].header.riff.format = HEADER_WAVE;
+	files[i].header.fmt.subchunk_id = HEADER_FMT;
+	files[i].header.fmt.subchunk_size = 16;
+	files[i].header.fmt.audio_format = 1;
+	files[i].header.fmt.num_channels = ((format & PROBE_MASK_NB_CHANNELS) >> PROBE_SHIFT_NB_CHANNELS) + 1;
+	files[i].header.fmt.sample_rate = sample_rate[(format & PROBE_MASK_SAMPLE_RATE) >> PROBE_SHIFT_SAMPLE_RATE];
+	files[i].header.fmt.bits_per_sample = (((format & PROBE_MASK_CONTAINER_SIZE) >> PROBE_SHIFT_CONTAINER_SIZE) + 1) * 8;
+	files[i].header.fmt.byte_rate = files[i].header.fmt.sample_rate *
+					files[i].header.fmt.num_channels *
+					files[i].header.fmt.bits_per_sample / 8;
+	files[i].header.fmt.block_align = files[i].header.fmt.num_channels *
+					  files[i].header.fmt.bits_per_sample / 8;
+	files[i].header.data.subchunk_id = HEADER_DATA;
+
+	fwrite(&files[i].header, sizeof(struct wave), 1, files[i].fd);
+
+	return i;
+}
+
+void finalize_wave_files(struct wave_files *files)
+{
+	uint32_t i, chunk_size;
+
+	for (i = 0; i < FILES_LIMIT; i++) {
+		if (files[i].fd) {
+			files[i].size *= 8;
+			chunk_size = files[i].size + 36;
+
+			fseek(files[i].fd, sizeof(uint32_t), SEEK_SET);
+			fwrite(&chunk_size, sizeof(uint32_t), 1, files[i].fd);
+			fseek(files[i].fd, sizeof(uint32_t) * 10, SEEK_SET);
+			fwrite(&files[i].size, sizeof(uint32_t), 1, files[i].fd);
+
+			fclose(files[i].fd);
+		}
+	}
+}
+
+int validate_data_packet(struct probe_data_packet *data_packet)
+{
+	uint32_t received_crc;
+	uint32_t calc_crc;
+
+	received_crc = data_packet->checksum;
+	data_packet->checksum = 0;
+	calc_crc = crc32(0, (char *)data_packet, sizeof(*data_packet));
+
+	if (received_crc == calc_crc)
+		return 1;
+	else
+		return 0;
+}
+
+void parse_data(char *file_in)
+{
+	FILE *fd_in;
+	struct wave_files files[FILES_LIMIT];
+	struct probe_data_packet *packet;
+	uint32_t data[DATA_READ_LIMIT];
+	uint32_t total_data_to_copy = 0;
+	uint32_t data_to_copy = 0;
+	uint32_t *w_ptr;
+
+	int i, j, file;
+
+	enum p_state state = 0;
+
+	fprintf(stdout, "%s:\t Parsing file: %s\n", APP_NAME, file_in);
+
+	fd_in = fopen(file_in, "rb");
+	if (!fd_in) {
+		fprintf(stderr, "error: unable to open file %s, error %d\n",
+			file_in, errno);
+		exit(0);
+	}
+
+	memset(&data, 0, sizeof(uint32_t) * DATA_READ_LIMIT);
+	memset(&files, 0, sizeof(struct wave_files) * FILES_LIMIT);
+	file = -1;
+
+	do {
+		i = fread(&data, sizeof(uint32_t), DATA_READ_LIMIT, fd_in);
+		for (j = 0; j < i; j++) {
+			if (data[j] == PROBE_EXTRACT_SYNC_WORD) {
+				if (state != READY) {
+					fprintf(stderr, "error: wrong state %d, err %d\n",
+						state, errno);
+					free(packet);
+					exit(0);
+				}
+				packet = malloc(sizeof(struct probe_data_packet));
```

Comment:
if we are only processing one packet at a time then we should allocate it once at init.

Reply:
This part needs commenting too.

Reply:
Not possible because we dont know how big will be this packet. We can for example at this stage have loaded only half of probe_data_packet so the size is unknown

Reply:
just allocate MAX packet size otherwise we have an extra kernel context switch for every packet

Reply:
Sorry, I misunderstood your first comment. One allocation at init is of course good idea, added with PACKET_MAX_SIZE